R/download_package.R

```diff
@@ -0,0 +1,327 @@
+                             check_first = TRUE) {
+    #' TODO: How many child levels should it support? -- currently one. 3 or 4 max
+    #' TODO: resource_map_pid argument accepts metadata pids - could change to not accept metadata pids
+    #' TODO: Add option for downloading in folder structure that mirrors nesting - rather than only all files in one folder
+    #' TODO: Convert check download size to helper function?
+
+    # Stop if the user doesn't have the pbapply package installed
+    if (!requireNamespace("pbapply")) {
+        stop(call. = FALSE,
+             "The pbapply package is required to show progress. Please install it and try again.")
+    }
+
+    # Check that input arguments are in the correct format
+    stopifnot(is.character(resource_map_pid))
+    stopifnot(is.character(download_directory))
+    stopifnot(file.exists(download_directory))
+    stopifnot(is.logical(check_download_size))
+    stopifnot(is.logical(download_child_packages))
+    stopifnot(is.logical(check_first))
+
+    # Get package pids
+    package <- arcticdatautils::get_package(mn, resource_map_pid, file_names = TRUE)
+    child_packages <- list()
+
+    # Get child package pids
+    if (download_child_packages == TRUE) {
+        # Check that child packages exist
+        if (length(package$child_packages) != 0) {
+            n_child_packages <- length(package$child_packages)
+            progressBar <- txtProgressBar(min = 0, max = n_child_packages, style = 3)
+
+            message("\nDownloading identifiers from child packages...")
+
+            # Loop through child packages and extract pids using get_package()
+            child_packages <- lapply(seq_len(n_child_packages), function(i) {
+                setTxtProgressBar(progressBar, i)
+                arcticdatautils::get_package(mn, package$child_packages[i], file_names = TRUE)
+            })
+
+            close(progressBar)
+        }
+    }
+
+    # Initialize data pids and filename prefixes vectors
+    data_pids <- vector("character")
+    filename_prefixes <- vector("character")
+
+    # Select data pids from initial package, if they exist
+    if (length(package$data) != 0) {
+        data_pids <- package$data
+    }
+
+    # Select data pids from child packages and add to data_pids
+    child_data_pids <- unlist(lapply(child_packages, function(package) {
+        return(package$data)
+    }))
+
+    data_pids <- c(data_pids, child_data_pids)
+
+    # Check that data exists
+    if (length(data_pids) == 0) {
+        stop("No data selected.  Double check the package you entered contains data files")
+    }
+
+    # Create filename prefixes
+    all_packages <- c(package, child_packages)
+    if (prefix_file_names == TRUE) {
+        filename_prefixes <- unlist(lapply(all_packages, function(package) {
+            prefix <- pid_to_prefix(all_packages$metadata)
+            return(rep(prefix, length(all_packages$data)))
+        }))
+    }
```

Comment:
unclear on this part. What does all_packages$metadata return?  Is this two metadata pids if you have child packages?  Does that mean everything uses the parent metadata PID as a prefix?

Reply:
This is unclear - added comments and a helper function to make it more readable